### PR TITLE
dev/core#3133 - Extend Authx functionality to support validation of e…

### PR DESCRIPTION
**Pull request by seamuslee001**
…xternally generated JWTs

Overview
----------------------------------------
This adds support for using externally generated JWTs e.g. by Auth0 to authenticate users using Authx

Before
----------------------------------------
Authx,when processing a Bearer token, expects the scope claim to be "authx" and the sub claim to be "cid:{thecontactid}", as generated by civicrm itself.

After
----------------------------------------
Other extensions can override the handling of checking of the credential in an example such as this extesnion https://github.com/australiangreens/civiauth0jwt

Technical Details
----------------------------------------
This changes to fire a new event `civi.authx.checkCredential` to allow extensions to validate if an authentication credential should be accepted or not. 

This is a re-submit of https://github.com/civicrm/civicrm-core/pull/23019 which some minor modifications to make it work with the current Authx code

### CODE REVIEW

File: ext/authx/README.md [hunk 1]
```diff
@@ -127,3 +129,83 @@ $ curl 'https://demouser:demopass@example.org/civicrm/authx/id'
 
 The "AuthX: Authenticate to services with password" CiviCRM permission must
 also be granted for the role associated to the user.
+
+## Extending credential handling
```
Comment (totten) [inline ext/authx/README.md:133]: Woot. Documentation. :)

At some point, we'd probably want to move this to dev-docs. But I'm mostly just glad it's written. :smile:

File: ext/authx/Civi/Authx/Authenticator.php [hunk 2]
```diff
@@ -64,6 +63,13 @@ public function on_civi_invoke_auth(GenericHookEvent $e) {
         _authx_redact(['_authx']);
       }
     }
+    // Accept legacy auth (?key=...&api_key=...) for 'civicrm/ajax/rest' and 'civicrm/ajax/api4/*'.
+    // The use of `?key=` could clash on some endpoints. Only accept on a small list of endpoints that are compatible with it.
+    if (count($e->args) > 2 && $e->args[1] === 'ajax' && in_array($e->args[2], ['rest', 'api4'])) {
+      if ((!empty($_REQUEST['api_key']) || !empty($_REQUEST['key']))) {
+        return (new LegacyRestAuthenticator())->auth($e, ['flow' => 'legacyrest', 'cred' => 'Bearer ' . $_REQUEST['api_key'] ?? '', 'siteKey' => $_REQUEST['key'] ?? NULL]);
```
Comment (totten) [inline ext/authx/Civi/Authx/Authenticator.php:70]: Haven't read/tested this branch, but I'm not sure you need to explicitly call out to `LegacyRestAuthenticator`. It should register itself.

Comment (seamuslee001) [reply ext/authx/Civi/Authx/Authenticator.php:70]: @totten yes I believe this was left over code from the previous version will remove

**Comment by seamuslee001** [comment]
I'm flagging this as merge ready as I'm treating myself as the reviewer here and also the code produced matches Tim's original draft of what he thought it should look like as per the comments on the previous PR, I'll give Tim a few more days if he wants to comment further but I understand the code and it seems fine to me

**Comment by colemanw** [comment]
@totten is this ready to merge?

**Comment by totten** [comment]
Yeah, I don't have any more to add, and I think it's had a decent pause. Let's merge it.

**Pull request merged**